File: c/linked_list.c
#include <stdio.h>
#include <stdlib.h>

struct linked_list {
    int value;
    struct linked_list* next;
};

void pushVal(struct linked_list* list_address, int value) {
    struct linked_list* ptr = list_address;

    while (ptr->next != NULL) {
        ptr = ptr->next;
    }

    ptr->next = (struct linked_list*) malloc(sizeof(struct linked_list));
    ptr->next->value = value;
    ptr->next->next = NULL;
}

void pushValues(struct linked_list* list_address, int values[], int arrayLength) {
    struct linked_list* ptr = list_address;
    for (int i = 0; i < arrayLength; i++) {
        while (ptr->next != NULL) {
            ptr = ptr->next;
        }

        ptr->next = (struct linked_list*) malloc(sizeof(struct linked_list));
        ptr->next->value = values[i];
        ptr->next->next = NULL;
    }
}

void print(struct linked_list* list_address) {
    struct linked_list* ptr = list_address;

    while (ptr != NULL) {
        printf("%d -> ", ptr->value);
        ptr = ptr->next;
    }
    (void) printf("NULL\n");
}

struct linked_list find(struct linked_list* list_address, int value) {
    struct linked_list* ptr = list_address;

    while (ptr->next != NULL) {
        if (ptr->value == value) {
            return *ptr;
        }
        ptr = ptr->next;
    }
    (void) printf("Could not find the value %d in the linked list.", value);
    // Return the original list if the value could not be found
    return *(list_address);
}

// Cannot insert to front
void insert(struct linked_list* list_address, int value, int index) {
    struct linked_list* ptr = list_address;
    struct linked_list* old;
    int i = 0;

    while (ptr->next != NULL) {
        if (i == index - 1) break;
        ptr = ptr->next;
        i++;
    }
    if (ptr->next == NULL) {
        ptr->next = (struct linked_list*) malloc(sizeof(struct linked_list));
        ptr->next->value = value;
        ptr->next->next = NULL;
    }
    else {
        old = ptr->next;
        ptr->next = (struct linked_list*) malloc(sizeof(struct linked_list));
        ptr->next->value = value;
        ptr->next->next = old;
    }
}

// Returns the new front of the list
struct linked_list* addToFront(struct linked_list* list_address, int value) {
    struct linked_list* ptr = (struct linked_list*) malloc(sizeof(struct linked_list));
    ptr->next = list_address;
    ptr->value = value;
    return ptr;
}

struct linked_list pop(struct linked_list* list_address, int index) {
    struct linked_list* ptr = list_address;
    struct linked_list* prev = NULL;
    int i = 0;
    while (ptr->next != NULL) {
        if (i == index) break;
        prev = ptr;
        ptr = ptr->next;
        i++;
    }
    struct linked_list removed;
    if (ptr->next != NULL) {
        prev->next = ptr->next;
        removed = *ptr;
        free(ptr);
    }
    return removed;
}

int main(int argc, char *argv[]) {
    struct linked_list* root = (struct linked_list*) malloc(sizeof(struct linked_list));
    root->value = 0;
    int values[] = {1, 2, 3, 4};
    pushVal(root, 10);
    print(root);
    pushValues(root, values, sizeof(values) / sizeof(values[0]));
    insert(root, 7, 1);
    root = addToFront(root, -1);
    print(root);
    (void) printf("Popped value: %d\n", pop(root, 3).value);
    print(root);
    return 0;
}

File: java/LinkedList.java
class LinkedList {
    public static class LinkedListObj {
        private double value;
        private LinkedListObj next;

        public LinkedListObj() {
            next = null;
        }

        public LinkedListObj(double value) {
            this.value = value;
            next = null;
        }

        public LinkedListObj(double value, LinkedListObj next) {
            this.value = value;
            this.next = next;
        }

        public void set(double value) {
            this.value = value;
        }

        // Add a new linked list object to the end of the array.
        public void append(double value) {
            LinkedListObj iter = this;
            while (iter.next != null) iter = iter.next;
            iter.next = new LinkedListObj(value);
        }

        public void append(double[] value) {
            LinkedListObj iter = this;
            while (iter.next != null) iter = iter.next;
            for (int i = 0; i < value.length; i++) {
                iter.next = new LinkedListObj(value[i]);
                iter = iter.next;
            }
        }

        // Add a new linked list object at index i (as perceived by this object).
        // If i is never reached, just append. i cannot equal 1.
        public void insert(double value, int i) {
            if (i == 0) addToFront(value);
            LinkedListObj iter = this;
            int counter = 0;
            while ((iter != null) && (counter != i - 1)) {
                iter = iter.next;
                counter++;
            }
            LinkedListObj temp = iter.next;
            iter.next = new LinkedListObj(value);
            iter.next.next = temp;
        }

        // Adds a new value to the front of the linked list and returns the new front
        public LinkedListObj addToFront(double value) {
            LinkedListObj newFront = new LinkedListObj(value, this);
            return newFront;
        }

        public double pop(int i) {
            LinkedListObj iter = next;
            LinkedListObj prev = this;
            int counter = 0;
            while ((iter != null) && (counter != i - 1)) {
                prev = iter;
                iter = iter.next;
                counter++;
            }
            if (iter == null) return -Double.MAX_VALUE;
            prev.next = iter.next;
            return iter.value;
        }

        public void print() {
            System.out.print(value + " -> ");
            LinkedListObj iter = next;
            while (iter != null) {
                System.out.print(iter.value + " -> ");
                iter = iter.next;
            }
            System.out.print("null\n");
        }
    }

    public static void main(String[] args) {
        LinkedListObj root = new LinkedListObj(0);
        root.append(new double[]{1, 1, 2, 3, 5, 8, 13, 21});
        root.print();
        root = root.addToFront(-1);
        root.print();
        System.out.println(root.pop(5));
        root.insert(10, 3);
        root.append(34);
        root.print();
    }
}